Make target pointer width in target json an integer

WaffleLapkin opened the pull request:
<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r? <reviewer name>
-->
<!-- homu-ignore:end -->
r? Noratrieb
cc @RalfJung (https://github.com/rust-lang/rust/pull/142352/files#r2230380120)



try-job: x86_64-rust-for-linux

WaffleLapkin committed:
turn pointer width into an integer in target.json

WaffleLapkin committed:
allow using `target_val!` with a rename

Noratrieb reviewed (commented):
I'm unsure if it's worth breaking everyone over such a minor improvement.. maybe? maybe not?

Noratrieb commented:
I'm inclined to say yes, but it would probably be good to get this merged today so it can be in the same nightly as the other breaking change instead of two consecutive nightlies

RalfJung commented:
> I'm unsure if it's worth breaking everyone over such a minor improvement.. maybe? maybe not?

We already did that with https://github.com/rust-lang/rust/pull/144443... the current state is just inconsistent.

But, no strong opinion.

WaffleLapkin commented:
@Noratrieb what's the other breaking change? #142352 has been merged for over a month..

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-19-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16522102805/job/46726154179) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/46726154179) [(plain)](https://github.com/rust-lang/rust/commit/267f0bd5233cfff5a30d9d71e0fe3008b5df6d7d/checks/46726154179/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[TIMING] core::build_steps::test::TestHelpers { target: aarch64-unknown-linux-gnu } -- 0.091
##[group]Testing stage2 compiletest suite=ui mode=ui (aarch64-unknown-linux-gnu)

thread 'main' panicked at src/tools/compiletest/src/common.rs:925:10:
called `Result::unwrap()` on an `Err` value: Error("invalid type: integer `64`, expected a string", line: 60, column: 30)
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Build completed unsuccessfully in 0:24:57
  local time: Fri Jul 25 13:07:23 UTC 2025
  network time: Fri, 25 Jul 2025 13:07:23 GMT
##[error]Process completed with exit code 1.
```

</details>

WaffleLapkin committed:
accept integer `target-pointer-width` in compiletest

WaffleLapkin commented:
The only relevant change I see is #144218, but that doesn't explain anything, it should have been failing with https://github.com/rust-lang/rust/pull/142352, I don't see how it could have accepted `c-int-width` being a string...

RalfJung commented:
It might be that the field of wrong type just got ignored, which is exactly why Nora ported this to serde.

WaffleLapkin commented on src/tools/compiletest/src/common.rs:
```diff
@@ -1064,7 +1065,8 @@ pub struct TargetCfg {
     pub(crate) families: Vec<String>,
-    #[serde(rename = "target-pointer-width", deserialize_with = "serde_parse_u32")]
+    #[serde(rename = "target-pointer-width")]
+    #[cfg_attr(bootstrap, serde(deserialize_with = "serde_parse_u32"))]
```
Apparently not!

A lot of `compiletest`'s self tests parse target json printed by `rustc`. Without this change `x t compiletest --stage 1` (or similarly `--stage 2`) fails. And if you remove the `cfg`-ed code parts completely, then `x t compiletest --stage 0` fails.

Both `x t compiletest --stage 0` and `x t compiletest --stage 2` are tested in CI, so this is the only way I've found so far that makes it work ^^'

RalfJung replied:
Oh, this will be a nightmare for future JSON format changes... but it is what it is, I guess.

Cc @jieyouxu

jieyouxu replied:
I'll look at this tmrw.

RalfJung replied:
@WaffleLapkin is it possible to make serde_parse_u32 accept both a string and an integer as input?

WaffleLapkin replied:
It is, although I'm concerned if we need such hacks for any change of the json format.

RalfJung replied:
Compiletest doesn't care about most of the format I think. But it does care about the pointer size.

RalfJung replied:
FWIW I still think making the parser robust to ints and strings here is the easiest solution. compiletest only needs a few fields and they won't change very often.

jieyouxu replied:
If changes in #144563 are merged, then we will only build and test `compiletest` against the stage >= 1 toolchains in CI, which should make `cfg(bootstrap)` unnecessary.

Noratrieb replied:
Relying on stage0 target info is simply wrong in case the target info changes between the versions (semantically), right? It should always use in tree target info.

RalfJung replied:
`x t compiletest --stage 0` runs the ui test with the beta (bootstrap) compiler, right?

No idea why that makes sense, but it means compiletest has to be able to parse the old and new format.

jieyouxu replied:
Discussing this in [#t-infra/bootstrap > Building vs testing &#96;compiletest&#96;](https://rust-lang.zulipchat.com/#narrow/channel/326414-t-infra.2Fbootstrap/topic/Building.20vs.20testing.20.60compiletest.60/with/531313874).

jieyouxu replied:
> #144675 should at least unblock you (I hope), there's a weird `RUSTC` read in `compiletest`'s own unit tests that I'll look into separately.

https://github.com/rust-lang/rust/pull/144675

WaffleLapkin commented on compiler/rustc_target/src/spec/mod.rs:
```diff
@@ -2311,7 +2311,7 @@ pub struct Target {
     /// Used for generating target documentation.
     pub metadata: TargetMetadata,
     /// Number of bits in a pointer. Influences the `target_pointer_width` `cfg` variable.
-    pub pointer_width: u32,
```
For consistency with `c-int-width` which is also `u16`.

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-19-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16654066925/job/47134403645) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/47134403645) [(plain)](https://github.com/rust-lang/rust/commit/c28a14b60f6f216cd8de7439ac603daea56ddb5f/checks/47134403645/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
---- [ui] tests/ui/check-cfg/values-target-json.rs stdout ----

error: auxiliary build of /checkout/tests/auxiliary/minicore.rs failed to compile: 
status: exit status: 1
command: "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/auxiliary/minicore.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/check-cfg/values-target-json/libminicore.rlib" "-A" "unused" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers" "--crate-type=lib" "--check-cfg=cfg()" "--target=/checkout/tests/ui/check-cfg/my-awesome-platform.json" "-Cpanic=abort" "-Cforce-unwind-tables=yes" "--crate-type" "rlib" "-Cpanic=abort"
stdout: none
--- stderr -------------------------------
error: error loading target specification: target-pointer-width: invalid type: string "64", expected u16 at line 6 column 32
  |
  = help: run `rustc --print target-list` for a list of built-in targets
------------------------------------------


---- [ui] tests/ui/codegen/mismatched-data-layouts.rs stdout ----

error: `build-fail` test is required to pass check build, but check build failed
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/codegen/mismatched-data-layouts.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/codegen/mismatched-data-layouts" "-A" "unused" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers" "--crate-type=lib" "--target=/checkout/tests/ui/codegen/mismatched-data-layout.json" "-Z" "unstable-options"
stdout: none
--- stderr -------------------------------
error: error loading target specification: target-pointer-width: invalid type: string "64", expected u16 at line 6 column 32
  |
  = help: run `rustc --print target-list` for a list of built-in targets
------------------------------------------


```

</details>

WaffleLapkin committed:
fix target-pointer-width in tests

bors commented:
:pushpin: Commit ffe4f0c5576829088fbb1c364e4d0551d9b20542 has been approved by `Noratrieb`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Noratrieb ffe4f0c5576829088fbb1c364e4d0551d9b20542 -->
<!-- homu: {"type":"Approved","sha":"ffe4f0c5576829088fbb1c364e4d0551d9b20542","approver":"Noratrieb","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

Zalathar commented:
Failed in rollup: https://github.com/rust-lang/rust/pull/144845#issuecomment-3146975152

@bors r-

Zalathar commented:
> I'm unsure if it's worth breaking everyone over such a minor improvement.. maybe? maybe not?

Well if nothing else we found a bunch of bootstrap/compiletest staging bugs. 😿

WaffleLapkin commented:
@bors r=Noratrieb

Zalathar commented:
Ah, the problem seems to be this cargo test including a hardcoded target.json file with the older syntax: https://github.com/rust-lang/cargo/pull/15026

WaffleLapkin commented:
I ran `rg '"target-pointer-width": "'` and fixed all the occurrences, maybe now it'll be fine :')

WaffleLapkin commented:
This is a problem in RFL, https://github.com/Rust-for-Linux/linux/blob/c17b750b3ad9f45f2b6f7e6f7f4679844244f0b9/scripts/generate_rust_target.rs#L228.

RFL is not a subtree/submodule, so I can't fix it in this PR. What is the policy for such cases?

WaffleLapkin commented:
@rustbot ping rfl

This PR changes `target-pointer-width` in the target spec to be an integer, rather than a string. You'll have to fix these two lines: [1](https://github.com/Rust-for-Linux/linux/blob/c17b750b3ad9f45f2b6f7e6f7f4679844244f0b9/scripts/generate_rust_target.rs#L228), [2](https://github.com/Rust-for-Linux/linux/blob/c17b750b3ad9f45f2b6f7e6f7f4679844244f0b9/scripts/generate_rust_target.rs#L248).

BennoLossin commented:
IIUC, this should fix the issue:
```patch
diff --git a/scripts/generate_rust_target.rs b/scripts/generate_rust_target.rs
index 39c82908ff3a..9c2e9a8fdbe0 100644
--- a/scripts/generate_rust_target.rs
+++ b/scripts/generate_rust_target.rs
@@ -225,7 +225,7 @@ fn main() {
         ts.push("features", features);
         ts.push("llvm-target", "x86_64-linux-gnu");
         ts.push("supported-sanitizers", ["kcfi", "kernel-address"]);
-        ts.push("target-pointer-width", "64");
+        ts.push("target-pointer-width", 64);
     } else if cfg.has("X86_32") {
         // This only works on UML, as i386 otherwise needs regparm support in rustc
         if !cfg.has("UML") {
@@ -245,7 +245,7 @@ fn main() {
         }
         ts.push("features", features);
         ts.push("llvm-target", "i386-unknown-linux-gnu");
-        ts.push("target-pointer-width", "32");
+        ts.push("target-pointer-width", 32);
     } else if cfg.has("LOONGARCH") {
         panic!("loongarch uses the builtin rustc loongarch64-unknown-none-softfloat target");
     } else {
```

IIRC from our discussions about the CI inclusion, there is a way for you to apply this patch on the CI until we get this fixed on our side.

To fix it permanently, we'd also need to change the behavior depending on the rust version, is this behind a feature gate, or is it just a change to a nightly feature? In the latter case, we'd need to check the rust version in that script (or have some other way to figure out what the target spec currently requires). I haven't looked into that, since I'm not too familiar with the specifics of our build system. Any suggestions from your side?

ojeda commented:
>IIUC, this should fix the issue:
 
 We also need to do it conditional to the compiler version, I suspect -- at least that is what I was going to do.

BennoLossin commented:
Ah classic Miguel, posts a comment right before I send mine :)

> We also need to do it conditional to the compiler version, I suspect -- at least that is what I was going to do.

Yeah I wrote about that later in my comment :)

ojeda commented:
Sorry... :)

> Yeah I wrote about that later in my comment :)

Yeah, `rustc_version_atleast` should do it -- I don't think it is a feature gate (but even if it were, it is likely easier to just do it conditional on the version).

ojeda committed:
CI: rfl: support Rust >= 1.91.0 target spec

To unblock the Rust CI in PR [1], use a temporary commit from Rust for
Linux that supports the future target spec format.

Link: https://github.com/rust-lang/rust/pull/144443 [1]
Signed-off-by: Miguel Ojeda <ojeda@kernel.org>

Noratrieb commented:
(btw you can pass try-jobs to bors try directly now :3)
@bors r+

WaffleLapkin commented:
https://github.com/rust-lang/rust/pull/145909 got merged, so this should be good to go. I can't imagine what else can break ^^'

WaffleLapkin commented:
@bors r=Noratrieb

bors commented:
:pushpin: Commit 2d8cb59784711f503170011461c23385042a974d has been approved by `Noratrieb`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Noratrieb 2d8cb59784711f503170011461c23385042a974d -->
<!-- homu: {"type":"Approved","sha":"2d8cb59784711f503170011461c23385042a974d","approver":"Noratrieb","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

WaffleLapkin commented:
yay :3
